src/nedb/indexes.js

```diff
@@ -0,0 +1,299 @@
+const BinarySearchTree = require('../binarySearchTree/avltree')
+const model = require('./model')
+const _ = require('underscore')
+const util = require('util')
+
+/**
+ * Two indexed pointers are equal iif they point to the same place
+ */
+function checkValueEquality (a, b) {
+  return a === b
+}
+
+/**
+ * Type-aware projection
+ */
+function projectForUnique (elt) {
+  if (elt === null) { return '$null' }
+  if (typeof elt === 'string') { return '$string' + elt }
+  if (typeof elt === 'boolean') { return '$boolean' + elt }
+  if (typeof elt === 'number') { return '$number' + elt }
+  if (util.isArray(elt)) { return '$date' + elt.getTime() }
+
+  return elt   // Arrays and objects, will check for pointer equality
+}
+
+class Index {
+  /**
+   * Create a new index
+   * All methods on an index guarantee that either the whole operation was successful and the index changed
+   * or the operation was unsuccessful and an error is thrown while the index is unchanged
+   * @param {String} options.fieldName On which field should the index apply (can use dot notation to index on sub fields)
+   * @param {Boolean} options.unique Optional, enforce a unique constraint (default: false)
+   * @param {Boolean} options.sparse Optional, allow a sparse index (we can have documents for which fieldName is undefined) (default: false)
+   */
+  constructor(options) {
+    this.fieldName = options.fieldName
+    this.unique = options.unique || false
+    this.sparse = options.sparse || false
+
+    this.treeOptions = { unique: this.unique, compareKeys: model.compareThings, checkValueEquality: checkValueEquality }
+
+    this.reset()   // No data in the beginning
+  }
+
+  /**
+   * Reset an index
+   * @param {Document or Array of documents} newData Optional, data to initialize the index with
+   *                                                 If an error is thrown during insertion, the index is not modified
+   */
+  reset(newData) {
+    this.tree = new BinarySearchTree(this.treeOptions)
+
+    if (newData) { this.insert(newData) }
+  }
+
+  /**
+   * Insert a new document in the index
+   * If an array is passed, we insert all its elements (if one insertion fails the index is not modified)
+   * O(log(n))
+   */
+  insert(doc) {
+    let key
+    const self = this
```

Comment:
looks like `self` is not used in this function at all, could be removed

Reply:
Some `self` variables can be replaced by using arrow functions